### ISSUE TITLE
Preemption for tibanna and azure

### ISSUE THREAD

**Issue by johanneskoester** [snakemake/snakemake#592]
Snakemake now has flags --preemption-default and --preemptible-rules, that allow to define which rules are eligible for being killed by the cloud provider (which leads to automatic restart). It would be great if the tibanna executor could use it, as well as the Azure executor (PR #533).

cc @andreas-wilm, @SooLee

**Comment by SooLee** [comment]
@johanneskoester Are these flags for preemptible instances (i.e. spot instances)? Tibanna fails the job when a spot instance interruption occurs. I assume then a Snakemake user could simply restart the workflow run and it would pick up from where it's left off? Did you have something different in mind?

**Comment by aryam7** [comment]
the related pr: #549

**Comment by andreas-wilm** [comment]
That's a very fruitful discussion in #549! I have to admit, I'm a bit on the fence with this one and if in doubt I very much favour simplicity (so I'd generally be for just using `--restart-times` as is). But I do see some value of being able to set lower restart values to long-running & expensive (in terms of cost) rules and higher values to long-running & cheap rules.

**Comment by gdeoli** [comment]
It would be really nice if tibanna could make use of something similar to --preemptible-rules. Currently, that's something I'm looking forward in some of my workflows, as I wish only some of the rules to be run in spot instances.

**Comment by SooLee** [comment]
Hi Gustavo,

Thank you for the message. Are there any specific rules that you're hoping
to use with spot instances? Currently Tibanna considers interruption as
failure and terminates the instance.

Best,
Soo

**Comment by gdeoli** [comment]
Yeah, it would be nice if I could choose which rules will go on spot or on-demand instances. Sometimes, rules have jobs that are relatively fast, thus I could make use of spot instances to save money. As it is today, my understanding is that tibanna will send all rules to be run on spot, if **--tibanna-config  spot_instance=true** is set. That's not the behavior I want at all, as some highly demanding rules, I know for sure, will be interrupted.

**Comment by SooLee** [comment]
I see. Thank you for the clarification. I could add something like
--spot-rules= which is equivalent to preemtible-rules. Preemtible is not
the term used by AWS, so spot may be a better name for the option. Thoughts?

**Comment by gdeoli** [comment]
That would be perfect! =)

**Comment by SichongP** [comment]
> I see. Thank you for the clarification. I could add something like --spot-rules= which is equivalent to preemtible-rules. Preemtible is not the term used by AWS, so spot may be a better name for the option. Thoughts?

To add to this, it would be great if we can specify in the snakemake files for each rule to be spot or on-demand, instead of listing everything via cli. An easy way without adding new directive is to use `nachine_type=spot/on_demand` in resources directive.

**Comment by cademirch** [comment]
Closing since this is no longer relevant here given the development of executor plugins.
